Add Motionblinds Bluetooth full test coverage

LennP opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Full test coverage for the Motionblinds Bluetooth integration

```bash
---------- coverage: platform linux, python 3.12.2-final-0 -----------
Name                                                       Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------
homeassistant/components/motionblinds_ble/__init__.py         52      0   100%
homeassistant/components/motionblinds_ble/button.py           28      0   100%
homeassistant/components/motionblinds_ble/config_flow.py     102      0   100%
homeassistant/components/motionblinds_ble/const.py            20      0   100%
homeassistant/components/motionblinds_ble/cover.py            89      0   100%
homeassistant/components/motionblinds_ble/entity.py           25      0   100%
homeassistant/components/motionblinds_ble/select.py           34      0   100%
homeassistant/components/motionblinds_ble/sensor.py           64      0   100%
----------------------------------------------------------------------------------------
TOTAL                                                        414      0   100%
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

LennP committed:
Add tests

LennP committed:
Fix entity test

LennP committed:
Format

LennP committed:
Fix sensor tests

LennP committed:
Fix sensor tests

LennP committed:
Fix sensor tests

LennP committed:
Add init tests

LennP committed:
Change service info

joostlek commented on tests/components/motionblinds_ble/__init__.py:
```diff
@@ -1 +1,61 @@
+)
+
+
+async def setup_platform(
+    hass: HomeAssistant,
+    platforms: list[Platform],
+    blind_type: MotionBlindType = MotionBlindType.ROLLER,
+) -> tuple[MockConfigEntry, str]:
+    """Mock a fully setup config entry."""
+
+    config_entry = MockConfigEntry(
+        title="mock_title",
+        domain=DOMAIN,
+        unique_id="cc:cc:cc:cc:cc:cc",
+        data={
+            CONF_ADDRESS: "cc:cc:cc:cc:cc:cc",
+            CONF_LOCAL_NAME: "Motionblind CCCC",
+            CONF_MAC_CODE: "CCCC",
+            CONF_BLIND_TYPE: blind_type.name.lower(),
+        },
+    )
+    config_entry.add_to_hass(hass)
+
+    with patch("homeassistant.components.motionblinds_ble.PLATFORMS", platforms):
+        assert await hass.config_entries.async_setup(config_entry.entry_id)
+        await async_setup_entry(hass, config_entry)
+        await hass.async_block_till_done()
+        assert config_entry.state is ConfigEntryState.LOADED
+
+    return (
+        config_entry,
+        str(config_entry.data[CONF_LOCAL_NAME]).lower().replace(" ", "_"),
+    )
```
I have mixed feelings about this function. I think we should better utilize mocks and fixtures and only patch the platforms when we really have a reason to.

LennP replied:
I guess the platforms patch can be removed, but I saw others use it so I included it as well (maybe for performance benefit?). The config_entry can be made a fixture, anything else?

joostlek replied:
I mean some integrations do it, but there is no real benefit at this point (like I have never seen anybody patch it for performance benefits, and if that was a thing, this integration would not be big enough for that). I would also remove the `assert config_entry.state is LOADED` because otherwise if you want to test a failing set up you need to copy paste everything

LennP replied:
Removed the assertion and the platform parameter. What would you use mocks or a fixture for here, considering the MockConfigEntry uses the blind_type parameter from the setup_platform function?

joostlek replied:
You can make the blind_type a fixture as well, you can overwrite that per test

LennP replied:
I don't understand what the advantage of that would be. You suggest having a fixture for the mock config entry with blind_type as parameter? How would you do that?

joostlek replied:
```py
@pytest.fixture()
def blind_type() -> MotionBlindType:
    return MotionBlindType.ROLLER

@pytest.fixture()
def mock_config_entry(blind_type: MotionBlindType) -> MockConfigEntry:
    return MockConfigEntry(
        data={
            CONF_BLIND_TYPE: blind_type.name.lower(),
        }
    )
```
You can then overwrite the blindtype per test via parametrize

LennP replied:
And then you would include mock_config_entry in the parameters of each test and pass it to setup_integration() explicitly?

> You can then overwrite the blindtype per test via parametrize

How?

joostlek replied:
Yep, or use a fixture that setup the integration. But I personally like the function more as it feels more flexible.

From the top of my head `@pytest.mark.parametrize("blind_type", ["a", "b"])`

LennP replied:
Since setup_integration returns a tuple I think it is better it remains a tuple indeed.

> From the top of my head @pytest.mark.parametrize("blind_type", ["a", "b"])

Sure but where and how would you use that to overwrite the default value for blind_type? For example this can use the default blind type:

```python
@pytest.mark.parametrize(("select", "args"), [(ATTR_SPEED, MotionSpeedLevel.HIGH)])
async def test_select(mock_config_entry, hass: HomeAssistant, select: str, args: Any) -> None:
    """Test select."""

    _, name = await setup_integration(hass, mock_config_entry)
```

But test_sensor requires a MotionBlindType.VENETIAN. How can I override the default value of blind_type that is used for mock_config_entry? Previously I just passed it as a kwarg: blind_type=MotionBlindType.CURTAIN

```python
# Parametrize here???
async def test_sensor(
    mock_config_entry,
    hass: HomeAssistant,
    sensor: str,
    update_func: Callable[[MotionDevice], Callable[..., None]],
    initial_value: str,
    args: list[Any],
    expected_value: str,
) -> None:
    """Test sensors."""

    config_entry, name = await setup_integration(
        hass, mock_config_entry # blind_type=MotionBlindType.CURTAIN
    )
    device: MotionDevice = hass.data[DOMAIN][config_entry.entry_id]
```

joostlek replied:
You can parametrize that test yes. You can als parametetrize multiple times on the same test

LennP replied:
Got it working by putting another parametrize above it

joostlek commented on tests/components/motionblinds_ble/__init__.py:
```diff
@@ -1 +1,61 @@
+
+    with patch("homeassistant.components.motionblinds_ble.PLATFORMS", platforms):
+        assert await hass.config_entries.async_setup(config_entry.entry_id)
+        await async_setup_entry(hass, config_entry)
```
This should not be called manually

LennP replied:
How else can the `async_setup_entry` be called, considering the integration includes both an `async_setup` and `async_setup_entry`?

joostlek replied:
the line above makes HA setup the integration. It will first call `async_setup` and then it will call `async_setup_entry`

LennP replied:
> the line above makes HA setup the integration. It will first call `async_setup` and then it will call `async_setup_entry`

Thing is I could not get the tests to work without the `await async_setup_entry(hass, config_entry)` line. Only using `await hass.config_entries.async_setup(config_entry.entry_id)` does not trigger `async_setup_entry` and therefore the platforms are not setup

joostlek replied:
That sounds like an issue on itself tbh

LennP replied:
Why would it be? Were async_setup and async_setup_entry not designed to work together? If so, why would you not be able to have a general integration setup and and a setup for each entry?

joostlek replied:
Well that's the thing, it should work together. when an entry is set up, `async_setup` is run, and then `async_setup_entry`. That's a given. If that is not the case, something is going wrong. Maybe something is not patched.

LennP replied:
Hmm not sure why. Without the `await async_setup_entry(hass, config_entry)` pretty much all tests fail as the platforms are not setup. I did check debug logs and the async_setup_entry is not called at all.

joostlek commented on tests/components/motionblinds_ble/test_button.py:
```diff
@@ -0,0 +1,36 @@
+
+    _, name = await setup_platform(hass, [Platform.BUTTON])
+
+    with patch(
+        f"homeassistant.components.motionblinds_ble.MotionDevice.{button}"
+    ) as command:
```
I think having a centralized mock is of a great benefit here

LennP replied:
You propose setting a Mock on each instance of the button entity? How would you do that?

joostlek replied:
No I mean a complete mock for `MotionDevice`. checkout `airgradient` for example. We create 1 mock for the client and we patch the methods that are needed and that;s it

LennP replied:
This has been solved!

joostlek commented on tests/components/motionblinds_ble/test_cover.py:
```diff
@@ -0,0 +1,113 @@
+        hass, [Platform.COVER], blind_type=MotionBlindType.VENETIAN
+    )
+
+    device: MotionDevice = hass.data[DOMAIN][config_entry.entry_id]
```
Don't touch internals in tests

LennP replied:
I can't access the MotionDevice?

joostlek replied:
Nope, mock everything and let the mocks play out the things you want to happen

LennP replied:
How should I do that then? Each entity registers callbacks on the MotionDevice when it is added. The MotionDevice is created in the async_setup_entry and then set in hass.data: `hass.data.setdefault(DOMAIN, {})[entry.entry_id] = device`. How do I make sure the device is the Mock in that case and therefore each entity registers a callback on the Mock?

joostlek replied:
Checkout `knocki` there I eventually fetch the callback function added to the client and call that to update states

LennP replied:
Do you know why using the mock_motion_device results in the unique_id changing from "cover.motionblind_cccc" (I guess that is the entry_id) to "cover.motionblinds_ble_cc_cc_cc_cc_cc_cc"

```
@pytest.mark.parametrize(
    ("running_type", "state"),
    [
        (None, "unknown"),
        (MotionRunningType.STILL, "unknown"),
        (MotionRunningType.OPENING, STATE_OPENING),
        (MotionRunningType.CLOSING, STATE_CLOSING),
    ],
)
async def test_cover_update_running(
    mock_config_entry: MockConfigEntry,
    mock_motion_device: Mock,  # Adding this line changes the unique_id
    hass: HomeAssistant,
    running_type: str | None,
    state: str,
) -> None:
    """Test updating running status."""

    name = await setup_integration(hass, mock_config_entry)

    async_update_running = mock_motion_device.register_running_callback.call_args[0][0]

    async_update_running(running_type)
    assert hass.states.get("cover.motionblinds_ble_cc_cc_cc_cc_cc_cc").state == state
    assert hass.states.get("cover.{name}").state == state
    # ^^ The last assertion suddenly doesn't work anymore since the unique_id changed when using mock_motion_device
```

joostlek replied:
Cover takes the device name and the device name is set to `device.display_name`. Is `display_name` properly mocked? Otherwise I think it will either follow the config entry title or the unique_id if `display_name` would suddently return nothing

LennP replied:
At this point I have no idea what is happening to the names. I have hardcoded some of the names in the code (see latest commits) to get 100% coverage but I am fully confused as to how and why these names are generated like this, and the logic required to get this working without hardcoding it:
- button.mock_title_connect
- button.disconnect
- cover.motionblinds_ble_cc_cc_cc_cc_cc_cc
- cover.motionblinds_cccc

Here are some of the states with all of these unique_ids:
[<state button.mock_title_connect=unknown; friendly_name=Connect @ 2024-07-13T10:07:24.430987-07:00>, <state button.disconnect=unknown; friendly_name=Disconnect @ 2024-07-13T10:07:24.432163-07:00>, <state button.favorite=unknown; friendly_name=Favorite @ 2024-07-13T10:07:24.432977-07:00>, <state cover.motionblinds_ble_cc_cc_cc_cc_cc_cc=unknown; device_class=shade, friendly_name=, supported_features=15 @ 2024-07-13T10:07:24.440620-07:00>, <state select.speed=unknown; options=['1', '2', '3'], friendly_name=Speed @ 2024-07-13T10:07:24.444602-07:00>, <state sensor.connection_status=disconnected; options=['connected', 'connecting', 'disconnected', 'disconnecting'], device_class=enum, friendly_name=Connection status @ 2024-07-13T10:07:24.449583-07:00>, <state sensor.signal_strength=unknown; unit_of_measurement=dBm, device_class=signal_strength, friendly_name=Signal strength @ 2024-07-13T10:07:24.451120-07:00>, <state sensor.battery=unknown; state_class=measurement, unit_of_measurement=%, device_class=battery, friendly_name=Battery @ 2024-07-13T10:07:24.451763-07:00>]

LennP replied:
You were right! display_name was indeed set to an empty string causing this odd behaviour. Fixed it now

LennP committed:
Rename test_sensor parameters

LennP committed:
Removce ConfigEntryState.LOADED assertion

LennP committed:
Remove platforms parameter from setup_platform

LennP committed:
Rename setup_platform to setup_integration

LennP committed:
Fixture for blind_type and mock_config_entry

LennP committed:
Use mock for MotionDevice

LennP committed:
Use mock for MotionDevice

LennP committed:
Add type hint

LennP committed:
Use Mock instead of patch

LennP committed:
Use mock_config_entry fixture

LennP committed:
Move constants to init

LennP committed:
Fix entity_id name

LennP committed:
Use fixture

joostlek commented on tests/components/motionblinds_ble/__init__.py:
```diff
@@ -1 +1,56 @@
+    await async_setup_entry(hass, mock_config_entry)
+    await hass.async_block_till_done()
+
+    return FIXTURE_DISPLAY_NAME.lower().replace(" ", "_")
```
I am also wondering why this function returns something static

LennP replied:
This has already been fixed by creating another fixture

joostlek commented on tests/components/motionblinds_ble/conftest.py:
```diff
@@ -3,11 +3,63 @@
+    """Mock a MotionDevice."""
+
+    with (
+        patch(
+            "homeassistant.components.motionblinds_ble.MotionDevice",
+            autospec=True,
+        ) as mock_device,
+        patch(
+            "homeassistant.components.motionblinds_ble.MotionDevice",
+            new=mock_device,
+        ),
```
This mock is the same. Doesnt the config flow also need a mock?

joostlek commented on tests/components/motionblinds_ble/test_button.py:
```diff
@@ -0,0 +1,48 @@
+
+    name = await setup_integration(hass, mock_config_entry)
+
+    command = AsyncMock()
+    setattr(mock_motion_device, button, command)
+
+    await hass.services.async_call(
+        BUTTON_DOMAIN,
+        SERVICE_PRESS,
+        {ATTR_ENTITY_ID: f"button.{name}_{button}"},
+        blocking=True,
+    )
+    command.assert_called_once()
```
```suggestion

    await hass.services.async_call(
        BUTTON_DOMAIN,
        SERVICE_PRESS,
        {ATTR_ENTITY_ID: f"button.{name}_{button}"},
        blocking=True,
    )
    getattr(mock_motion_device, command).assert_called_once()
```

joostlek commented on tests/components/motionblinds_ble/test_cover.py:
```diff
@@ -0,0 +1,123 @@
+
+    name = await setup_integration(hass, mock_config_entry)
+
+    with patch(
+        f"homeassistant.components.motionblinds_ble.MotionDevice.{method}"
+    ) as func:
```
Can also use the other way of patching

LennP replied:
Correct done

joostlek commented on tests/components/motionblinds_ble/test_entity.py:
```diff
@@ -0,0 +1,60 @@
+from tests.common import MockConfigEntry
+
+
+@pytest.mark.parametrize(
+    ("platform", "entity"),
+    [
+        (Platform.COVER, None),
+        (Platform.BUTTON, ATTR_CONNECT),
+        (Platform.BUTTON, ATTR_DISCONNECT),
+        (Platform.BUTTON, ATTR_FAVORITE),
+        (Platform.SELECT, ATTR_SPEED),
+    ],
+)
+async def test_entity_update(
```
This is a platform concern, so this would rather fit a test in each platform

LennP replied:
That would only result in duplicate code? Why not just test it here that every entity responds to the update_entity service? I could also just move it to test_cover.py and only test the cover as that will still result in 100% coverage but this is better imo

joostlek replied:
Hmm, the second thing is that we don't like `if` statements in the code because it makes it more unclear what its testing. So I would maybe prefer only the cover.

I am also low key wondering, why is this needed for the 100%? What is missing otherwise?

LennP replied:
> I am also low key wondering, why is this needed for the 100%? What is missing otherwise?

The `async_update` method of the MotionblindsBLEEntity class.

```python
    async def async_update(self) -> None:
        """Update state, called by HA if there is a poll interval and by the service homeassistant.update_entity."""
        _LOGGER.debug("(%s) Updating entity", self.entry.data[CONF_MAC_CODE])
        await self.device.status_query()
```

> Hmm, the second thing is that we don't like if statements in the code because it makes it more unclear what its testing. So I would maybe prefer only the cover.

In my opinion the added parametrize and condition are worth it to cover all platforms and make sure each of them can be used to update the device state (after all they are all an instance of MotionblindsBLEEntity), but if you insist on only doing it for the cover platform let me know and I will change it

edenhaus replied:
In my opinion, this test is OK and reduces duplicated code except for the cover variant. We should avoid if statements in tests, and therefore please split the cover variant of this test into a new test, placed in test_cover.py

LennP replied:
I do not see the appeal of duplicating this exact same code for the test_cover.py over using a simple if statement that determines the entity_id. I could also create a get_unique_id() method that constructs the unique_id based on the platform and type of entity such that the if statement is no longer in the test? For now though I have removed the cover platform as it will reach 100% coverage either way as long as one platform is tested

LennP committed:
Use fixtures instead of constants

LennP committed:
Use display_name fixture

LennP committed:
Rename mac to mac_code

LennP committed:
Remove one patch

LennP committed:
Use fixtures for mock_config_entry

LennP committed:
Apply suggestion

LennP committed:
Replace patch with mock

joostlek commented:
I'll have to check that code out tomorrow because that async setup entry is a dealbreaker and should not be there. So maybe there's something else going on, but I'll have to check

LennP committed:
Replace patch with mock

LennP committed:
Replace patch with mock

LennP commented:
> I'll have to check that code out tomorrow because that async setup entry is a dealbreaker and should not be there. So maybe there's something else going on, but I'll have to check

Sure! I'm not sure how else to fix it as it does not work without it. Let me know!

joostlek committed:
Fix

joostlek commented:
You had the `async_setup_entry` patched in another test that was marked with autouse=True.

I think there is some more cleanup to do, because you marked it with autouse=True, but you also have `@pytest.mark.usefixtures("motionblinds_ble_connect")` at every test, so either one of those two ways to include the fixture should go

LennP committed:
Use pytest.mark.usefixtures if parameter not used

LennP commented:
Ah thanks! I had not noticed that as I haven't created that test. I removed autouse=True and replaced a few unused mock_setup_entry parameters with pytest.mark.usefixtures

LennP committed:
Base mac code on address

LennP committed:
Remove if statement from entity test
